File: frontend/src/types/playType.ts
export interface PlayTypeMetrics {
  possessions: number | null
  points: number | null
  ppp: number | null
  fg_pct: number | null
  frequency: number | null
  ppp_percentile: number | null
}

export interface SpotUpMetrics extends PlayTypeMetrics {
  fg3m: number | null
  fg3a: number | null
  fg3_pct: number | null
}

export interface PlayerPlayTypeStats {
  id: number
  nba_id: number
  name: string
  position: string | null
  team_abbreviation: string | null
  season: string
  total_poss: number | null
  isolation: PlayTypeMetrics | null
  pnr_ball_handler: PlayTypeMetrics | null
  pnr_roll_man: PlayTypeMetrics | null
  post_up: PlayTypeMetrics | null
  spot_up: SpotUpMetrics | null
  transition: PlayTypeMetrics | null
  cut: PlayTypeMetrics | null
  off_screen: PlayTypeMetrics | null
}

export interface PlayTypeLeaderboardEntry {
  rank: number | null
  id: number
  nba_id: number
  name: string
  position: string | null
  team_abbreviation: string | null
  possessions: number | null
  points: number | null
  ppp: number | null
  fg_pct: number | null
  frequency: number | null
  ppp_percentile: number | null
}

export interface PlayTypeLeaderboardResponse {
  play_type: string
  sort_by: string
  entries: PlayTypeLeaderboardEntry[]
}

export type PlayTypeKey =
  | 'isolation'
  | 'pnr_ball_handler'
  | 'pnr_roll_man'
  | 'post_up'
  | 'spot_up'
  | 'transition'
  | 'cut'
  | 'off_screen'

export type PlayTypeSortBy = 'ppp' | 'possessions' | 'fg_pct' | 'frequency'

export const PLAY_TYPE_LABELS: Record<PlayTypeKey, string> = {
  isolation: 'Isolation',
  pnr_ball_handler: 'PnR Ball Handler',
  pnr_roll_man: 'PnR Roll Man',
  post_up: 'Post-Up',
  spot_up: 'Spot-Up',
  transition: 'Transition',
  cut: 'Cut',
  off_screen: 'Off Screen',
}

export const SORT_BY_LABELS: Record<PlayTypeSortBy, string> = {
  ppp: 'PPP',
  possessions: 'Possessions',
  fg_pct: 'FG%',
  frequency: 'Frequency',
}

File: backend/app/schemas/play_type.py
"""Pydantic schemas for play type statistics."""

from decimal import Decimal

from pydantic import BaseModel, ConfigDict


class PlayTypeMetrics(BaseModel):
    """Metrics for a single play type."""

    possessions: int | None
    points: int | None
    ppp: Decimal | None  # Points per possession
    fg_pct: Decimal | None  # Field goal percentage
    frequency: Decimal | None  # Frequency of this play type
    ppp_percentile: int | None = None


class SpotUpMetrics(PlayTypeMetrics):
    """Metrics for spot-up plays (includes 3-point tracking)."""

    fg3m: int | None = None
    fg3a: int | None = None
    fg3_pct: Decimal | None = None


class PlayerPlayTypeStats(BaseModel):
    """Full play type stats for a player."""

    id: int
    nba_id: int
    name: str
    position: str | None
    team_abbreviation: str | None
    season: str
    total_poss: int | None

    isolation: PlayTypeMetrics | None = None
    pnr_ball_handler: PlayTypeMetrics | None = None
    pnr_roll_man: PlayTypeMetrics | None = None
    post_up: PlayTypeMetrics | None = None
    spot_up: SpotUpMetrics | None = None
    transition: PlayTypeMetrics | None = None
    cut: PlayTypeMetrics | None = None
    off_screen: PlayTypeMetrics | None = None

    model_config = ConfigDict(from_attributes=True)


class PlayTypeLeaderboardEntry(BaseModel):
    """Entry in the play type leaderboard."""

    rank: int | None = None
    id: int
    nba_id: int
    name: str
    position: str | None
    team_abbreviation: str | None
    possessions: int | None
    points: int | None
    ppp: Decimal | None
    fg_pct: Decimal | None
    frequency: Decimal | None
    ppp_percentile: int | None = None

    model_config = ConfigDict(from_attributes=True)


class PlayTypeLeaderboardResponse(BaseModel):
    """Response for play type leaderboard endpoint."""

    play_type: str
    sort_by: str
    entries: list[PlayTypeLeaderboardEntry]
